bumble/profiles/hap.py

```diff
@@ -0,0 +1,599 @@
+
+
+@dataclass
+class PresetRecord:
+    '''See Hearing Access Service 2.8. Preset record.'''
+
+    @dataclass
+    class Property:
+        class Writable(OpenIntEnum):
+            CANNOT_BE_WRITTEN = 0b0
+            CAN_BE_WRITTEN = 0b1
+
+        class IsAvailable(OpenIntEnum):
+            IS_UNAVAILABLE = 0b0
+            IS_AVAILABLE = 0b1
+
+        writable: Writable = Writable.CAN_BE_WRITTEN
+        is_available: IsAvailable = IsAvailable.IS_AVAILABLE
+
+        def __bytes__(self) -> bytes:
+            return bytes([self.writable | (self.is_available << 1)])
+
+    index: int
+    name: str
+    properties: Property = field(default_factory=Property)
+
+    def __bytes__(self) -> bytes:
+        return bytes([self.index]) + bytes(self.properties) + self.name.encode('utf-8')
+
+    def is_available(self) -> bool:
+        return (
+            self.properties.is_available
+            == PresetRecord.Property.IsAvailable.IS_AVAILABLE
+        )
+
+
+# -----------------------------------------------------------------------------
+# Server
+# -----------------------------------------------------------------------------
+class HearingAccessService(gatt.TemplateService):
+    UUID = gatt.GATT_HEARING_ACCESS_SERVICE
+
+    hearing_aid_features_characteristic: gatt.Characteristic
+    hearing_aid_preset_control_point: gatt.Characteristic
+    active_preset_index_characteristic: gatt.Characteristic
+    active_preset_index: int = 0x00
+    # TODO validate address type → public_address ?
+    active_preset_index_per_device: Dict[Address, int] = {}
+
+    device: Device
+
+    server_features: HearingAidFeatures
+    preset_records: Dict[int, PresetRecord] = {}  # key is the preset index
+    read_presets_request_in_progress: bool = False
+
+    preset_changed_operations_history_per_device: Dict[
+        Address, List[PresetChangedOperation]
+    ] = {}
+
+    # Keep an updated list of connected client to send notification to
+    currently_connected_clients: Set[Connection] = set()
+
+    def __init__(
+        self, device: Device, features: HearingAidFeatures, presets: List[PresetRecord]
+    ) -> None:
+        # 2.5. GATT sub-procedure requirements
+        assert att.ATT_DEFAULT_MTU >= 49
+
+        self.device = device
+        self.server_features = features
+        for p in presets:
+            assert len(p.name.encode()) >= 1
+            assert len(p.name.encode()) <= 40
+            self.preset_records[p.index] = p
+
+        @device.on('connection')
+        def on_connection(connection: Connection) -> None:
+            @connection.on('disconnection')
+            def on_disconnection(_reason):
+                self.currently_connected_clients.remove(connection)
+            # TODO Should we filter on device bonded && device is HAP ?
+            self.currently_connected_clients.add(connection)
+            if connection.device.public_address not in self.preset_changed_operations_history_per_device:
```

Comment:
Is the address must be public?

Reply:
I do not know
HAS is only involving bonded device, so I believe public_address is the one we want to use ?

Reply:
Device might distribute random static address as its identity address, an then the statement here will not be correct.
I think we may have a logic similar to SMP https://github.com/google/bumble/blob/main/bumble/smp.py#L1080 to properly choose a proper identity address. (We may make it a method provided by device.py when Gilles is back)

Reply:
BTW, it seems like you are always putting `connection.device.public_address` which is the address of local device (not remote), is that really what you want? The address of remote device is provided by `connection.peer_address`

Reply:
I do want the remote device address.
I have updated the code to use `peer_address`

HAS is only exposing service that require encryption, I understand that mean the device have to be bonded and thus the peer_address would be correct.
Let me know if I am wrong.